### DISCUSSION TITLE
Can anyone explain how generateStaticParams() works along with the param object in next 13?

### DISCUSSION THREAD

**Discussion by alenart91** [vercel/next.js#44641] [Help]
From what I understand the generateStaticParams() function runs at build time and uses the returned list to generate the static pages. So it seems like it executes the code of the dynamic page.js file (for instance blog/[slug]/page.js) for every slug and outputs an html file in the build folder with whatever the slug is. I don't understand how the params object works. It looks like it uses some kind of routing object to match the url to the slug?

**Comment by icyJoseph** [comment]
It is very similar to `getStaticPaths`, so `generateStaticParams`  has to return every path member that's dynamic. 

In `/app/[foo]/[bar]/page.tsx`, `generateStaticParams` has to return a list of objects, where the keys are `foo` and `bar`, and the values are strings, or list of strings for catch-all paths.

For example from the docs:

```js
// app/products/[categorySlug]/[productId]/page.js
// Generate segments for both [categorySlug] and [productId]
export async function generateStaticParams() {
  const rows = await query('SELECT category, id FROM products');

  return rows.map((row) => ({
    categorySlug: row.category,
    productId: row.id,
  }));
}
```

Does that help out?

**Reply by nouman-stackone** [reply]
Thanks for explaining but I'm having it tough to understand. Could you please give a real-world simple example?

**Reply by icyJoseph** [reply]
If you have a book store, with various types of books, `history`, `novels`, etc, and each book has an `id`. Then you'd call your `getAllBooks` API and create combinations `/books/[type]/[book-id]`.

This will generate a page for each type + id combination, that is all of your books.

The `params` object in your page would have access to the type and id, for any logic you'd need within that context.

When using `getStaticPaths` this would also be passed to the `getStaticProps` to get the data, but since Server Components can do that themselves, `generateStaticPaths` passes it directly to the server component.

For example:

```tsx
import { notFound } from "next/navigation";

type Book = {
  type: string;
  id: string;
  price: number;
  title: string;
  author: string;
};

const db: Record<string, Book> = {
  "000": {
    type: "novel",
    id: "000",
    price: 10,
    title: "hello",
    author: "joe doe",
  },
  "001": {
    type: "novel",
    id: "001",
    price: 1,
    title: "fishing",
    author: "anon",
  },
  "002": {
    type: "sci-fi",
    id: "002",
    price: 100,
    title: "42",
    author: "jane doe",
  },
};

const getBookById = async (id: string) => {
  return Promise.resolve(db[id]);
};

const BookPage = async ({ params }: { params: Pick<Book, "type" | "id"> }) => {
  const book = await getBookById(params.id);

  if (!book) {
    notFound();
  }

  return (
    <article>
      <header>{book.title}</header>
    </article>
  );
};

export async function generateStaticParams() {
  const books = await Promise.resolve(Object.values(db));

  return books.map(({ type, id }) => ({
    type,
    id,
  }));
}

export default BookPage;
```

And when we build the project:

```console
Route (app)                                Size     First Load JS
┌ λ /                                      0 B                0 B
└ ● /books/[type]/[id]                     185 B          82.5 kB
    ├ /books/novel/000
    ├ /books/novel/001
    └ /books/sci-fi/002
+ First Load JS shared by all              82.3 kB
  ├ chunks/17-4e8fdc7dead168ed.js          80.4 kB
  ├ chunks/main-app-3ca53ed61af9674f.js    216 B
  └ chunks/webpack-23701e5bacce31e3.js     1.65 kB
```

With an `appDir` folder structure:

```console
app
├── books
│   └── [type]
│       └── [id]
│           └── page.tsx
├── error.tsx
├── head.tsx
├── layout.tsx
├── loading.tsx
└── page.tsx
```

**Reply by 17AtGitHub** [reply]
Can you also please share how is it helpful to use generateStaticParams? When a request is made to a UI route at say 
`localhost:3000/books/sci-fi/001`, what is the sequence of steps followed by the application to display the UI?
When it makes the first req to the server, does the server generate all dynamic routes and store in an array? And when some other UI component is requested with a similar route, the server can directly get the API from the array instead of again making request to get all the Books collection, and look for data related to the bookID?
Is this how it is useful? 

Kindly share if, at any place, I am stating wrong technically. I am willing to understand all this, but I am new to this field. 
Thanks.

**Reply by nomadscode** [reply]
Hi @icyJoseph, thanks for the example. I tried using it because I kept encountering an **undefined** ID in my own project. However, after running your code, I still end up with an **undefined** ID. It's neither possible for me to run a build of the pages, those will simply not be build. Are you aware of any changes in Next.js that might be causing this issue? Cheers, Bart

[A sandbox with the above code](https://codesandbox.io/p/sandbox/long-bird-2r3q7q)

**Reply by O4epegb** [reply]
@bartfluitsma I've answered you on SO https://stackoverflow.com/a/76599630/5650447

tl;dr: You have mismatch between dynamic segment name and what you expect in the code. The naming should be consistent between folder name and what you expect in the code. If segment (folder) is named `[type]` then you should return `type` key, you cannot name it `bookType` or something like that, it should match the folder name.

**Reply by icyJoseph** [reply]
@bartfluitsma your path is `/books/[type]/[book-id]`, you need to do:

```ts
export async function generateStaticParams() {
  const books = await Promise.resolve(Object.values(db));

  return books.map(({ type, id }) => ({
    type,
    'book-id': id,
  }));
}
```

And then read it from params:

```tsx
// the params type annotation is not correct it needs type and 'book-id' :) 
const BookPage = async ({ params }: { params: Pick<Book, "type" | "id"> }) => {
  const book = await getBookById(params['book-id']);
```

I leave fixing the `params` type annotation up to you, as homework.

**Reply by nomadscode** [reply]
Thank you @O4epegb, for both replies! Thank you @icyJoseph for your help (and the homework ;) )!  It works now.

I'm really curious, from my understanding, the data is fetched based on the given ID, the `book-id` in this case. Which also makes up for the URL example _/books/novel/001_. But what if I would like to have the URL based on the `book title`, but still fetch the data via the `book-id`, how would that work?

**Reply by icyJoseph** [reply]
You'd need to implement a getByTitle function to read your data source. Typically people use a "slug" instead of the title. This means not using the id at all...

Some also append the id to the slug, but often that's because they can't implement the getByTitle. 

You could also have a triple param path, type/title/id :) 

Also if you wanna do getByTitle, the title better be unique.

**Reply by nomadscode** [reply]
Ah, that's too bad! I'm fetching data from Google Firebase, and need the ID to fetch the corresponding data, but it's an ugly string not great for SEO/sharing friendly urls - and if I use the name or slug, I can't fetch the data like I do with the ID. 

Thanks for your help and the tips @icyJoseph , appreciate it!

**Reply by chukwukap** [reply]
Can we use it inside the app router

**Reply by mertafor** [reply]
There is one "deadly" important thing about generateStaticParams which is it generates all pages on build time rather than runtime. In other words, it generates prebuilt html files that cannot access any server-side dynamic params such as headers and cookies. Otherwise you'll get "Dynamic server usage" error. This function should be better used for particular paths rather than entire app (layout.tsx) in most cases. Yesterday I wasted my entire day because this function suggested to be used for internalization in the docs. Then I found the reason totally by chance in one of the discussions here. I know that's also my carelesness yet I hope Next team highlights this vital information wherever this function is suggested to be used. Otherwise, an improved error message pointing out that  generateStaticParams is involved would be really appreciated.

**Reply by Mintnoii** [reply]
I encountered the same problem when developing with the notion database api, how did you solve it?

**Reply by Julymycin** [reply]
Can I know how you get the Route structure?

**Reply by mercurymirror** [reply]
I have a problem with generateStaticParams because I can't find the files in the out folder after the build. Am I missing something or a step?

**Reply by hectorzea** [reply]
your comment helped me a lot and i think is the most important one here to take a look @mertafor

i was trying a POC using **react-i18next**  in the most actual nextjs version with app router  and i faced this error.

In the official tutorial of [react-i18next](https://github.com/i18next/next-i18next) they use **generateStaticParams** at layout level (check the documentation [here](https://www.locize.com/blog/i18n-next-app-router) )
in layout.tsx they used a getT function which used 'next/headers' and that can generate some errors if you want to generate the page at build time.

i know for sure that if pages are statically generated site could be more faster, but in an scenario like this i dont know if its worth to find different workarounds.

**Reply by mertafor** [reply]
@hectorzea I remember I've had to build my own i18n solution back then. But afterwards I started using [next-intl](https://next-intl.dev/). 

I don't know if there are any better i18n libraries for app router out there but this library is in active development since the early days of Next 14 beta release and by far the best for dealing with RSC and static rendering limitations. Also it's listed in official Next JS docs.

**Comment by gabriel-logan** [comment]
You can take a look at my repository, I'm using an example of generateStaticParams very clearly there, you can make a clone and take a look at the code to understand it better, it will help you.

https://github.com/gabriel-logan/generateStaticParams-nextjs14

Basically generateStaticParams works as follows, when you use dynamic routes, nextjs can create these routes based on the parameter that is passed to the browser, however when you use a static page, the page cannot update the information dynamically.

When you use generateStaticParams

nextjs at build time "generates" all the parameters that are possible, IN THE BOTTOM actually what happens is that nextjs creates an html for EACH route you wanted to access so you need to pass the endpoint where ALL the ids are .

In my example you can see, I'm looking for data from an API so I pass it into my staticParams then when the next build the application it makes a request to the address that I put there and returns the array containing the number of items within the array , so when you access the parameters on the static page, it will already have ALL the pages preloaded.

Basically, static pages are not good when you are dealing with routes because after the build is done IF a new item appears in the API it will not update as it is only generated at build time.

**Reply by JazzyMichael** [reply]
> Basically, static pages are not good when you are dealing with routes because after the build is done IF a new item appears in the API it will not update as it is only generated at build time.

This is not accurate. What you described is a basic form of static site generation that is completely different from nextjs' incremental static generation offering. ISR is intended to generate static pages at runtime. If the revalidation time has passed, the first visitor gets a stale version of the page while a new page is statically generated. The only static pages generated at build time are ones that are specified in `generateStaticParams`, however returning an empty array should force generation of dynamic routes at runtime.

From the docs: https://nextjs.org/docs/app/api-reference/functions/generate-static-params "You must return [an empty array from generateStaticParams](https://nextjs.org/docs/app/api-reference/functions/generate-static-params#all-paths-at-build-time) or utilize [export const dynamic = 'force-static'](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamic) in order to revalidate (ISR) [paths at runtime](https://nextjs.org/docs/app/api-reference/functions/generate-static-params#all-paths-at-runtime). "

**Comment by istealersn-dev** [comment]
@gabriel-logan You seem to be making fetch call twice isn't it? Once inside `generateStaticParams` and Post function. I am struggling to understand why does Nextjs require us to make fetch call twice in order to prebuild HTML page?

The only thing I can understand is Nextjs builds the dynamicroute pages based on `generateStaticParams()` however, what happens if you attempt a Post that doesn't exist? My expectation would be that it returns 404 rather it doesn't instead it renders the last known page. Once I added ` export const dynamicParams = false` it started returning 404

As an experiment, I excluded from passing params to the Page function to see if my page renders fine and to my surprise it seem to LOAD fine. Apparently, page doesn't need the params unless you are going to make use of `slug`. Does it mean if a new post is created at the server where API is hosted it will not render on request? as I am not passing the params

I would love some more understanding

**Reply by gabriel-logan** [reply]
You need to pass it twice because the fetch you do inside generateStaticParams is theoretically responsible for "creating the parameters" that will be accessed in the static routes.

Generate an html File for each route

 And yes, if you pass a post that does not exist at the time of build, it will always return 404 because.

 When you generate static pages in next

 It generates all the files that are accessible in the build, so as I said if new data appears in the api it won't find it, you will have to generate a new build.

**Reply by designly1** [reply]
Next.js (or React actually, I think) extends the fetch API and does cool stuff like automatic deduplicating :D. I used to wonder the same thing but pretty cool eh?

**Reply by shweta-codeprism** [reply]
@gabriel-logan have you tried this combination:
```
export const dynamicParams = true
export const revalidate = [n]
```

This should revalidate the page/layout every n seconds. So unless you have a different cache time for fetching the params, the params should get updated every n seconds. No need to rebuild.

**Comment by shamun-khatri** [comment]
@icyJoseph I have the following code below but when I build that is not going to build and it throws error.


```
import React from "react";
import ProductSlider from "@/components/ProductSlider";
import { getProducts } from "@/utils/api";
import Banner from "@/components/Banner";
import "./page.css";
import { notFound } from "next/navigation";

// export const revalidate = 1;
export const revalidate = 1;
export const dynamicParams = true;

const fetchStoreData = async (username) => {
  const res = await fetch(`http://localhost:3000/api/user/${username}`);
  const data = await res.json();
  return data;
};

const UserProfile = async ({ params }) => {

  const data = await fetchStoreData(params.username);

  if (data.userInfo.length === 0) {
    return notFound();
  }

  const products = await getProducts();

  return (
    <div>
      <Banner />
      <ProductSlider products={products} />
      <h1>test man</h1>
      <h1>User Profile: {data.userInfo[0].email}</h1>
    </div>
  );
};

export async function generateStaticParams() {
  const storeNames = await fetch(
    `http://localhost:3000/api/user/store/getStoreNames`, { next: { revalidate: 1 } }
  ).then((res) => res.json());

  const stores = [...storeNames.storeNames];


  return stores.map((store) => ({ username: store }));
}

export default UserProfile;
```

how can I generate a static page for every user? or store

**Reply by icyJoseph** [reply]
You are trying to fetch from your own server at build time. More here: https://nextjs-faq.com/fetch-api-in-rsc

In short, just get the data you need, in the generate static params function, there is no need to make an extra request toward your own, non existing during build time, server.

**Reply by shamun-khatri** [reply]
@icyJoseph Yes, I think you are right, But how can I solve this.

**Reply by icyJoseph** [reply]
Move the contents of `api/user/[username]/route.ts`, or API route equivalent, into the get user function, and adjust the inputs you need. Also bring any imports necessary, and read the FAQ link I posted.

**Comment by aafnnp** [comment]


**Reply by aafnnp** [reply]
https://github.com/aafnnp/site
